File: app/src/main/java/com/arjanvlek/oxygenupdater/settings/BottomSheetPreference.java
package com.arjanvlek.oxygenupdater.settings;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.res.TypedArrayUtils;
import androidx.preference.Preference;

import com.arjanvlek.oxygenupdater.R;
import com.google.android.material.bottomsheet.BottomSheetDialog;

import java.util.ArrayList;
import java.util.List;

import static android.view.View.GONE;
import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;
import static com.arjanvlek.oxygenupdater.internal.logger.Logger.logInfo;
import static com.arjanvlek.oxygenupdater.internal.logger.Logger.logVerbose;
import static com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_EXPANDED;
import static java.lang.Math.min;
import static java.lang.String.format;
import static java.util.Locale.getDefault;

/**
 * Overridden class to add custom functionality (setting title/message, custom on click listeners for dismissing).
 *
 * @author Adhiraj Singh Chauhan (github.com/adhirajsinghchauhan)
 */
@SuppressWarnings("WeakerAccess")
public class BottomSheetPreference extends Preference {

	private static final String TAG = "BottomSheetPreference";

	private Context context;
	private OnPreferenceChangeListener mOnChangeListener;

	private LinearLayout dialogLayout;
	private LinearLayout itemListContainer;
	private BottomSheetDialog dialog;

	private String secondaryKey;
	private String title;
	private String caption;
	private List<BottomSheetItem> itemList = new ArrayList<>();
	private boolean valueSet;
	private Object value;
	private Object secondaryValue;

	private SettingsManager settingsManager;

	public BottomSheetPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		super(context, attrs, defStyleAttr, defStyleRes);

		init(context, attrs, defStyleAttr, defStyleRes);
	}

	public BottomSheetPreference(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);

		init(context, attrs, defStyleAttr, 0);
	}

	public BottomSheetPreference(Context context, AttributeSet attrs) {
		super(context, attrs);

		init(context, attrs, 0, 0);
	}

	public BottomSheetPreference(Context context) {
		super(context);

		init(context, null, 0, 0);
	}

	/**
	 * Initialise the preference
	 *
	 * @param context      the context
	 * @param attrs        the attributes
	 * @param defStyleAttr default style attributes
	 * @param defStyleRes  default style resource
	 */
	private void init(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		this.context = context;

		settingsManager = new SettingsManager(context);

		readAttrs(attrs, defStyleAttr, defStyleRes);

		if (secondaryKey == null) {
			secondaryKey = getKey() + "_id";
		}

		setupDialog();
	}

	/**
	 * Setup the internal {@link BottomSheetDialog}
	 */
	@SuppressLint("InflateParams")
	private void setupDialog() {
		dialog = new BottomSheetDialog(context);
		LayoutInflater inflater = LayoutInflater.from(context);

		dialogLayout = (LinearLayout) inflater.inflate(R.layout.bottom_sheet, null, false);
		itemListContainer = dialogLayout.findViewById(R.id.dialog_item_list_container);

		setText(dialogLayout.findViewById(R.id.dialog_title), title);
		setText(dialogLayout.findViewById(R.id.dialog_caption), caption);

		logVerbose(TAG, format(getDefault(), "Setup dialog with title='%s', subtitle='%s', and '%d' items", title, caption, itemList.size()));

		for (int i = 0; i < itemList.size(); i++) {
			LinearLayout dialogItemLayout = (LinearLayout) inflater.inflate(R.layout.bottom_sheet_item, itemListContainer, false);

			// add the item's view at the specified index
			itemListContainer.addView(dialogItemLayout, i);

			setupItemView(dialogItemLayout, i);
		}

		dialog.setContentView(dialogLayout);

		// Open up the dialog fully by default
		dialog.getBehavior().setState(STATE_EXPANDED);
	}

	/**
	 * Initialises an item layout. Sets text if needed, and marks selected/unselected as well
	 *
	 * @param dialogItemLayout parent linear layout
	 * @param index            item index
	 */
	private void setupItemView(LinearLayout dialogItemLayout, int index) {
		BottomSheetItem item = itemList.get(index);

		TextView titleView = dialogItemLayout.findViewById(R.id.dialog_item_title);
		TextView subtitleView = dialogItemLayout.findViewById(R.id.dialog_item_subtitle);

		setText(titleView, item.getTitle());
		setText(subtitleView, item.getSubtitle());

		logVerbose(TAG, format(getDefault(), "Setup item with title='%s', and subtitle='%s'", title, caption));

		dialogItemLayout.setOnClickListener(view -> setValueIndex(index));

		Object currentValue = settingsManager.getPreference(getKey(), null);
		Object currentSecondaryValue = settingsManager.getPreference(secondaryKey, null);

		Object secondaryValue = item.getSecondaryValue();

		// value is mandatory, secondary value is optional
		if (item.getValue().equals(currentValue) || (secondaryValue != null && secondaryValue.equals(currentSecondaryValue))) {
			markItemSelected(index);
		} else {
			markItemUnselected(index);
		}
	}

	private void redrawItemView(int index) {
		LinearLayout dialogItemLayout = (LinearLayout) itemListContainer.getChildAt(index);

		setupItemView(dialogItemLayout, index);
	}

	/**
	 * Reads attributes defined in XML and sets relevant fields
	 *
	 * @param attrs        the attributes
	 * @param defStyleAttr default style attribute
	 * @param defStyleRes  default style resource
	 */
	private void readAttrs(AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.BottomSheetPreference, defStyleAttr, defStyleRes);

		title = TypedArrayUtils.getString(a, R.styleable.BottomSheetPreference_title, R.styleable.BottomSheetPreference_android_title);
		caption = a.getString(R.styleable.BottomSheetPreference_caption);
		secondaryKey = a.getString(R.styleable.BottomSheetPreference_secondaryKey);

		CharSequence[] entries = a.getTextArray(R.styleable.BottomSheetPreference_android_entries);
		CharSequence[] entryValues = a.getTextArray(R.styleable.BottomSheetPreference_android_entryValues);

		CharSequence[] titleEntries = a.getTextArray(R.styleable.BottomSheetPreference_titleEntries);
		CharSequence[] subtitleEntries = a.getTextArray(R.styleable.BottomSheetPreference_subtitleEntries);
		int resourceId = a.getResourceId(R.styleable.BottomSheetPreference_secondaryEntryValues, 0);

		itemList = new ArrayList<>();

		int[] intEntryValues = resourceId != 0
				? context.getResources().getIntArray(resourceId)
				: null;

		int numberOfItems = entries != null
				? entries.length
				: titleEntries != null
				? titleEntries.length
				: 0;

		for (int i = 0; i < numberOfItems; i++) {
			BottomSheetItem item = BottomSheetItem.builder()
					.value(entryValues[i].toString())
					.build();

			CharSequence title = entries != null ? entries[i] : titleEntries[i];
			CharSequence subtitle = subtitleEntries != null ? subtitleEntries[i] : null;
			Object secondaryValue = intEntryValues != null ? intEntryValues[i] : null;

			if (title != null) {
				item.setTitle(title.toString());
			}

			if (subtitle != null) {
				item.setSubtitle(subtitle.toString());
			}

			if (secondaryValue != null) {
				item.setSecondaryValue(secondaryValue);
			}

			itemList.add(item);
		}

		a.recycle();
	}

	/**
	 * Utility method that sets text if the supplied text is not null, otherwise hides the {@link TextView}
	 *
	 * @param textView the TextView
	 * @param text     the text
	 */
	private void setText(TextView textView, String text) {
		if (text != null) {
			textView.setVisibility(VISIBLE);
			textView.setText(text);
		} else {
			textView.setVisibility(GONE);
		}
	}

	/**
	 * Finds index of item with newValue or newSecondaryValue and calls {@link #setValueIndex(int)}
	 *
	 * @param newValue          the new value
	 * @param newSecondaryValue the new integer value
	 */
	private void setValues(Object newValue, Object newSecondaryValue) {
		int selectedIndex = -1;

		for (int i = 0; i < itemList.size(); i++) {
			BottomSheetItem item = itemList.get(i);
			Object value = item.getValue();
			Object secondaryValue = item.getSecondaryValue();

			if ((value != null && value.equals(newValue)) || (secondaryValue != null && secondaryValue.equals(newSecondaryValue))) {
				selectedIndex = i;
				break;
			}
		}

		if (selectedIndex != -1) {
			setValueIndex(selectedIndex);
		}
	}

	/**
	 * Sets the value of the key. secondaryValue is optional.
	 * <p>
	 * Also marks previous item as unselected and new item as selected
	 *
	 * @param selectedIndex the selected index
	 */
	public void setValueIndex(int selectedIndex) {
		BottomSheetItem item = itemList.get(selectedIndex);

		Object newValue = item.getValue();
		Object newSecondaryValue = item.getSecondaryValue();

		// Always persist/notify the first time.
		boolean changed = !newValue.equals(value);

		if (changed || !valueSet) {
			value = newValue;
			secondaryValue = newSecondaryValue;
			valueSet = true;

			settingsManager.savePreference(getKey(), newValue);

			if (newSecondaryValue != null) {
				settingsManager.savePreference(secondaryKey, newSecondaryValue);
			}

			if (changed) {
				// redraw previous selected and new selected layouts
				remarkItems(newValue);
				onChange();
			}
		}
	}

	public void setSecondaryKey(String secondaryKey) {
		logInfo(TAG, format(getDefault(), "Updating secondaryKey: %s", secondaryKey));

		this.secondaryKey = secondaryKey;
	}

	@Override
	public void setTitle(CharSequence title) {
		logInfo(TAG, format(getDefault(), "Updating dialog title: %s", title));

		this.title = title.toString();

		setText(dialogLayout.findViewById(R.id.dialog_title), title.toString());

		super.setTitle(title);
	}

	@Override
	public void onClick() {
		if (dialog != null) {
			dialog.show();
		}
	}

	/**
	 * Sets the callback to be invoked when this preference is changed by the user (but before
	 * the internal state has been updated).
	 *
	 * @param onPreferenceChangeListener The callback to be invoked
	 */
	@Override
	public void setOnPreferenceChangeListener(OnPreferenceChangeListener onPreferenceChangeListener) {
		mOnChangeListener = onPreferenceChangeListener;
		super.setOnPreferenceChangeListener(onPreferenceChangeListener);
	}

	@Override
	protected void onSetInitialValue(Object defaultValue) {
		Object newValue = settingsManager.getPreference(getKey(), null);
		Object newSecondaryValue = settingsManager.getPreference(secondaryKey, null);

		setValues(newValue, newSecondaryValue);
	}

	@Override
	protected Parcelable onSaveInstanceState() {
		Parcelable superState = super.onSaveInstanceState();
		if (isPersistent()) {
			// No need to save instance state since it's persistent
			return superState;
		}

		SavedState myState = new SavedState(superState);
		myState.value = value;
		myState.secondaryValue = secondaryValue;
		return myState;
	}

	@Override
	protected void onRestoreInstanceState(Parcelable state) {
		if (state == null || !state.getClass().equals(SavedState.class)) {
			// Didn't save state for us in onSaveInstanceState
			super.onRestoreInstanceState(state);
			return;
		}

		SavedState myState = (SavedState) state;
		super.onRestoreInstanceState(myState.getSuperState());

		setValues(myState.value, myState.secondaryValue);
	}

	public void setCaption(String caption) {
		logInfo(TAG, format(getDefault(), "Updating dialog caption: %s", caption));

		this.caption = caption;

		setText(dialogLayout.findViewById(R.id.dialog_caption), caption);
	}

	public void setEntries(CharSequence[] entries) {
		logInfo(TAG, format(getDefault(), "Updating titles for %d items", entries.length));

		// don't care if items and entries aren't equal in length
		// replace title for all indices that are common
		int length = min(itemList.size(), entries.length);

		for (int i = 0; i < length; i++) {
			itemList.get(i).setTitle(entries[i].toString());

			redrawItemView(i);
		}

		onSetInitialValue(null);
	}

	public void setTitleEntries(CharSequence[] titles) {
		logInfo(TAG, format(getDefault(), "Updating titles for %d items", titles.length));

		// don't care if items and titles aren't equal in length
		// replace title for all indices that are common
		int length = min(itemList.size(), titles.length);

		for (int i = 0; i < length; i++) {
			itemList.get(i).setTitle(titles[i].toString());

			redrawItemView(i);
		}

		onSetInitialValue(null);
	}

	public void setSubtitleEntries(CharSequence[] subtitles) {
		logInfo(TAG, format(getDefault(), "Updating subtitles for %d items", subtitles.length));

		// don't care if items and subtitles aren't equal in length
		// replace subtitle for all indices that are common
		int length = min(itemList.size(), subtitles.length);

		for (int i = 0; i < length; i++) {
			itemList.get(i).setSubtitle(subtitles[i].toString());

			redrawItemView(i);
		}
	}

	public void setEntryValues(CharSequence[] entryValues) {
		logInfo(TAG, format(getDefault(), "Updating entryValues for %d items", entryValues.length));

		// don't care if items and entryValues aren't equal in length
		// replace value for all indices that are common
		int length = min(itemList.size(), entryValues.length);

		for (int i = 0; i < length; i++) {
			itemList.get(i).setValue(entryValues[i].toString());
		}
	}

	public void setSecondaryEntryValues(Object[] objectEntryValues) {
		logInfo(TAG, format(getDefault(), "Updating secondaryValues for %d items", objectEntryValues.length));

		// don't care if items and objectEntryValues aren't equal in length
		// replace secondaryValue for all indices that are common
		int length = min(itemList.size(), objectEntryValues.length);

		for (int i = 0; i < length; i++) {
			itemList.get(i).setSecondaryValue(objectEntryValues[i]);
		}
	}

	public void setItemList(@NonNull List<BottomSheetItem> itemList) {
		logInfo(TAG, format(getDefault(), "Populating itemList with %d items", itemList.size()));

		this.itemList = new ArrayList<>();
		// copy list instead of shallow-referencing it
		this.itemList.addAll(itemList);

		setupDialog();

		onSetInitialValue(null);
	}

	public Object getValue() {
		return value;
	}

	public Object getSecondaryValue() {
		return secondaryValue;
	}

	/**
	 * Iterates over {@link #itemList} to remark items as selected/unselected
	 *
	 * @param value the new value
	 */
	private void remarkItems(Object value) {
		logVerbose(TAG, "Remarking items as selected/unselected");

		for (int i = 0; i < itemList.size(); i++) {
			BottomSheetItem item = itemList.get(i);

			if (item.getValue().equals(value)) {
				markItemSelected(i);
			} else {
				markItemUnselected(i);
			}
		}
	}

	/**
	 * Hides the checkmark icon and resets background to {@link android.R.attr#selectableItemBackground}
	 *
	 * @param selectedIndex index of the selected item
	 */
	private void markItemUnselected(int selectedIndex) {
		if (selectedIndex != -1) {
			LinearLayout dialogItemLayout = (LinearLayout) itemListContainer.getChildAt(selectedIndex);
			ImageView checkmarkView = dialogItemLayout.findViewById(R.id.dialog_item_checkmark);

			checkmarkView.setVisibility(INVISIBLE);
			TypedValue outValue = new TypedValue();
			context.getTheme().resolveAttribute(android.R.attr.selectableItemBackground, outValue, true);
			dialogItemLayout.setBackgroundResource(outValue.resourceId);

			logVerbose(TAG, format(getDefault(), "Item #%d marked unselected with title='%s', and subtitle='%s'", selectedIndex, title, caption));
		}
	}

	/**
	 * Shows the checkmark icon and sets background to {@link R.drawable#rounded_overlay}
	 *
	 * @param selectedIndex index of the selected item
	 */
	private void markItemSelected(int selectedIndex) {
		if (selectedIndex != -1) {
			LinearLayout dialogItemLayout = (LinearLayout) itemListContainer.getChildAt(selectedIndex);
			ImageView checkmarkView = dialogItemLayout.findViewById(R.id.dialog_item_checkmark);

			checkmarkView.setVisibility(VISIBLE);
			dialogItemLayout.setBackgroundResource(R.drawable.rounded_overlay);

			logVerbose(TAG, format(getDefault(), "Item #%d marked selected with title='%s', and subtitle='%s'", selectedIndex, title, caption));
		}
	}

	/**
	 * Called when value changes. Updates summary, closes the {@link #dialog}, an notifies on change listeners
	 */
	private void onChange() {
		setSummary(String.valueOf(value));

		dialog.cancel();

		if (mOnChangeListener != null) {
			mOnChangeListener.onPreferenceChange(this, value);
		}
	}

	private static class SavedState extends BaseSavedState {
		public static final Parcelable.Creator<SavedState> CREATOR =
				new Parcelable.Creator<SavedState>() {
					@Override
					public SavedState createFromParcel(Parcel in) {
						return new SavedState(in);
					}

					@Override
					public SavedState[] newArray(int size) {
						return new SavedState[size];
					}
				};

		Object value;
		Object secondaryValue;

		SavedState(Parcel source) {
			super(source);
			value = source.readString();
			secondaryValue = source.readLong();
		}

		SavedState(Parcelable superState) {
			super(superState);
		}

		@Override
		public void writeToParcel(@NonNull Parcel dest, int flags) {
			super.writeToParcel(dest, flags);
			dest.writeString(String.valueOf(value));
		}
	}
}

File: app/src/main/java/com/arjanvlek/oxygenupdater/settings/BottomSheetPreference.kt
package com.arjanvlek.oxygenupdater.settings

import android.annotation.SuppressLint
import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View.GONE
import android.view.View.INVISIBLE
import android.view.View.VISIBLE
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.res.TypedArrayUtils
import androidx.preference.Preference
import com.arjanvlek.oxygenupdater.R
import com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_EXPANDED
import com.google.android.material.bottomsheet.BottomSheetDialog
import java.lang.Math.min
import java.util.*

/**
 * Overridden class to add custom functionality (setting title/message, custom on click listeners for dismissing).
 *
 * @author Adhiraj Singh Chauhan (github.com/adhirajsinghchauhan)
 */
class BottomSheetPreference : Preference {

    private var contextRef: Context? = null
    private var mOnChangeListener: OnPreferenceChangeListener? = null

    private var dialogLayout: LinearLayout? = null
    private var itemListContainer: LinearLayout? = null
    private var dialog: BottomSheetDialog? = null

    private var secondaryKey: String? = null
    private var title: String? = null
    private var caption: String? = null
    private var itemList: MutableList<BottomSheetItem> = ArrayList()
    private var valueSet: Boolean = false
    var value: String? = null
        private set
    var secondaryValue: Any? = null
        private set

    private var settingsManager: SettingsManager? = null

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {

        init(context, attrs, defStyleAttr, defStyleRes)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {

        init(context, attrs, defStyleAttr, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {

        init(context, attrs, 0, 0)
    }

    constructor(context: Context) : super(context) {

        init(context, null, 0, 0)
    }

    /**
     * Initialise the preference
     *
     * @param context      the contextRef
     * @param attrs        the attributes
     * @param defStyleAttr default style attributes
     * @param defStyleRes  default style resource
     */
    private fun init(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) {
        this.contextRef = context

        secondaryKey = key + "_id"
        settingsManager = SettingsManager(context)

        readAttrs(attrs, defStyleAttr, defStyleRes)

        setupDialog()
    }

    /**
     * Setup the internal [BottomSheetDialog]
     */
    @SuppressLint("InflateParams")
    private fun setupDialog() {
        dialog = BottomSheetDialog(contextRef!!)
        val inflater = LayoutInflater.from(contextRef)

        dialogLayout = inflater.inflate(R.layout.bottom_sheet, null, false) as LinearLayout
        itemListContainer = dialogLayout!!.findViewById(R.id.dialog_item_list_container)

        setText(dialogLayout!!.findViewById(R.id.dialog_title), title)
        setText(dialogLayout!!.findViewById(R.id.dialog_caption), caption)

        for (i in itemList.indices) {
            val dialogItemLayout = inflater.inflate(R.layout.bottom_sheet_item, itemListContainer, false) as LinearLayout

            // add the item's view at the specified index
            itemListContainer!!.addView(dialogItemLayout, i)

            setupItemView(dialogItemLayout, i)
        }

        dialog!!.setContentView(dialogLayout)

        // Open up the dialog fully by default
        dialog!!.behavior.state = STATE_EXPANDED
    }

    /**
     * Initialises an item layout. Sets text if needed, and marks selected/unselected as well
     *
     * @param dialogItemLayout parent linear layout
     * @param index            item index
     */
    private fun setupItemView(dialogItemLayout: LinearLayout, index: Int) {
        val item = itemList[index]

        val titleView = dialogItemLayout.findViewById<TextView>(R.id.dialog_item_title)
        val subtitleView = dialogItemLayout.findViewById<TextView>(R.id.dialog_item_subtitle)

        setText(titleView, item.title)
        setText(subtitleView, item.subtitle)

        dialogItemLayout.setOnClickListener { setValueIndex(index) }

        val currentValue = settingsManager!!.getPreference(key, "")

        if (item.value == currentValue) {
            markItemSelected(index)
        } else {
            markItemUnselected(index)
        }
    }

    private fun redrawItemView(index: Int) {
        val dialogItemLayout = itemListContainer!!.getChildAt(index) as LinearLayout

        setupItemView(dialogItemLayout, index)
    }

    /**
     * Reads attributes defined in XML and sets relevant fields
     *
     * @param attrs        the attributes
     * @param defStyleAttr default style attribute
     * @param defStyleRes  default style resource
     */
    private fun readAttrs(attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) {
        val a = contextRef!!.obtainStyledAttributes(attrs, R.styleable.BottomSheetPreference,
                defStyleAttr, defStyleRes)

        title = TypedArrayUtils.getString(a, R.styleable.BottomSheetPreference_title,
                R.styleable.BottomSheetPreference_android_title)
        caption = a.getString(R.styleable.BottomSheetPreference_caption)

        val entries = a.getTextArray(R.styleable.BottomSheetPreference_android_entries)
        val entryValues = a.getTextArray(R.styleable.BottomSheetPreference_android_entryValues)

        val titleEntries = a.getTextArray(R.styleable.BottomSheetPreference_titleEntries)
        val subtitleEntries = a.getTextArray(R.styleable.BottomSheetPreference_subtitleEntries)
        val resourceId = a.getResourceId(R.styleable.BottomSheetPreference_secondaryEntryValues, 0)

        itemList = ArrayList()

        val intEntryValues = if (resourceId != 0)
            contextRef!!.resources.getIntArray(resourceId)
        else
            null

        val numberOfItems = entries?.size ?: (titleEntries?.size ?: 0)

        for (i in 0 until numberOfItems) {
            val item = BottomSheetItem(value = entryValues[i].toString())

            val title = if (entries != null) entries[i] else titleEntries!![i]
            val subtitle = if (subtitleEntries != null) subtitleEntries[i] else null
            val secondaryValue = if (intEntryValues != null) intEntryValues[i].toLong() else null

            if (title != null) {
                item.title = title.toString()
            }

            if (subtitle != null) {
                item.subtitle = subtitle.toString()
            }

            if (secondaryValue != null) {
                item.secondaryValue = secondaryValue
            }

            itemList.add(item)
        }

        a.recycle()
    }

    /**
     * Utility method that sets text if the supplied text is not null, otherwise hides the [TextView]
     *
     * @param textView the TextView
     * @param text     the text
     */
    private fun setText(textView: TextView, text: String?) {
        if (text != null) {
            textView.visibility = VISIBLE
            textView.text = text
        } else {
            textView.visibility = GONE
        }
    }

    /**
     * Finds index of item with newValue or newSecondaryValue and calls [.setValueIndex]
     *
     * @param newValue          the new value
     * @param newSecondaryValue the new integer value
     */
    private fun setValues(newValue: String?, newSecondaryValue: Any?) {
        var selectedIndex = -1

        for (i in itemList.indices) {
            val item = itemList[i]
            val value = item.value
            val secondaryValue = item.secondaryValue

            if (value != null && value == newValue || secondaryValue != null && secondaryValue == newSecondaryValue) {
                selectedIndex = i
                break
            }
        }

        if (selectedIndex != -1) {
            setValueIndex(selectedIndex)
        }
    }

    /**
     * Sets the value of the key. secondaryValue is optional.
     *
     *
     * Also marks previous item as unselected and new item as selected
     *
     * @param selectedIndex the selected index
     */
    fun setValueIndex(selectedIndex: Int) {
        val item = itemList[selectedIndex]

        val newValue = item.value
        val newSecondaryValue = item.secondaryValue

        // Always persist/notify the first time.
        val changed = !TextUtils.equals(value, newValue)

        if (changed || !valueSet) {
            value = newValue
            secondaryValue = newSecondaryValue
            valueSet = true

            persistString(newValue)

            if (newSecondaryValue != null) {
                settingsManager!!.savePreference(secondaryKey!!, newSecondaryValue)
            }

            if (changed) {
                // redraw previous selected and new selected layouts
                remarkItems(newValue)
                onChange()
            }
        }
    }

    fun setSecondaryKey(secondaryKey: String) {
        this.secondaryKey = secondaryKey
    }

    override fun setTitle(title: CharSequence) {
        this.title = title.toString()

        setText(dialogLayout!!.findViewById(R.id.dialog_title), title.toString())

        super.setTitle(title)
    }

    public override fun onClick() {
        if (dialog != null) {
            dialog!!.show()
        }
    }

    /**
     * Sets the callback to be invoked when this preference is changed by the user (but before
     * the internal state has been updated).
     *
     * @param onPreferenceChangeListener The callback to be invoked
     */
    override fun setOnPreferenceChangeListener(onPreferenceChangeListener: OnPreferenceChangeListener) {
        mOnChangeListener = onPreferenceChangeListener
        super.setOnPreferenceChangeListener(onPreferenceChangeListener)
    }

    override fun onSetInitialValue(defaultValue: Any?) {
        val newValue = getPersistedString(defaultValue as String?)
        val newSecondaryValue = settingsManager!!.getPreference<Any>(secondaryKey!!, "")

        setValues(newValue, newSecondaryValue)
    }

    override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        if (isPersistent) {
            // No need to save instance state since it's persistent
            return superState
        }

        val myState = SavedState(superState)
        myState.value = value
        myState.secondaryValue = secondaryValue
        return myState
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        if (state == null || state.javaClass != SavedState::class.java) {
            // Didn't save state for us in onSaveInstanceState
            super.onRestoreInstanceState(state)
            return
        }

        val myState = state as SavedState?
        super.onRestoreInstanceState(myState!!.superState)

        setValues(myState.value, myState.secondaryValue)
    }

    fun setCaption(caption: String) {
        this.caption = caption

        setText(dialogLayout!!.findViewById(R.id.dialog_caption), caption)
    }

    fun setEntries(entries: Array<CharSequence>) {
        // don't care if items and entries aren't equal in length
        // replace title for all indices that are common
        val length = min(itemList.size, entries.size)

        for (i in 0 until length) {
            itemList[i].title = entries[i].toString()

            redrawItemView(i)
        }

        onSetInitialValue(null)
    }

    fun setTitleEntries(titles: Array<CharSequence>) {
        // don't care if items and titles aren't equal in length
        // replace title for all indices that are common
        val length = min(itemList.size, titles.size)

        for (i in 0 until length) {
            itemList[i].title = titles[i].toString()

            redrawItemView(i)
        }

        onSetInitialValue(null)
    }

    fun setSubtitleEntries(subtitles: Array<CharSequence>) {
        // don't care if items and subtitles aren't equal in length
        // replace subtitle for all indices that are common
        val length = min(itemList.size, subtitles.size)

        for (i in 0 until length) {
            itemList[i].subtitle = subtitles[i].toString()

            redrawItemView(i)
        }
    }

    fun setEntryValues(entryValues: Array<CharSequence>) {
        // don't care if items and entryValues aren't equal in length
        // replace value for all indices that are common
        val length = min(itemList.size, entryValues.size)

        for (i in 0 until length) {
            itemList[i].value = entryValues[i].toString()
        }
    }

    fun setSecondaryEntryValues(objectEntryValues: Array<Any>) {
        // don't care if items and objectEntryValues aren't equal in length
        // replace secondaryValue for all indices that are common
        val length = min(itemList.size, objectEntryValues.size)

        for (i in 0 until length) {
            itemList[i].secondaryValue = objectEntryValues[i]
        }
    }

    fun setItemList(itemList: List<BottomSheetItem>) {
        this.itemList = ArrayList()
        // copy list instead of shallow-referencing it
        this.itemList.addAll(itemList)

        setupDialog()

        onSetInitialValue(null)
    }

    /**
     * Iterates over [.itemList] to remark items as selected/unselected
     *
     * @param value the new value
     */
    private fun remarkItems(value: String?) {
        for (i in itemList.indices) {
            val item = itemList[i]

            if (item.value == value) {
                markItemSelected(i)
            } else {
                markItemUnselected(i)
            }
        }
    }

    /**
     * Hides the checkmark icon and resets background to [android.R.attr.selectableItemBackground]
     *
     * @param selectedIndex index of the selected item
     */
    private fun markItemUnselected(selectedIndex: Int) {
        if (selectedIndex != -1) {
            val dialogItemLayout = itemListContainer!!.getChildAt(selectedIndex) as LinearLayout
            val checkmarkView = dialogItemLayout.findViewById<ImageView>(R.id.dialog_item_checkmark)

            checkmarkView.visibility = INVISIBLE
            val outValue = TypedValue()
            contextRef!!.theme.resolveAttribute(android.R.attr.selectableItemBackground, outValue, true)
            dialogItemLayout.setBackgroundResource(outValue.resourceId)
        }
    }

    /**
     * Shows the checkmark icon and sets background to [R.drawable.rounded_overlay]
     *
     * @param selectedIndex index of the selected item
     */
    private fun markItemSelected(selectedIndex: Int) {
        if (selectedIndex != -1) {
            val dialogItemLayout = itemListContainer!!.getChildAt(selectedIndex) as LinearLayout
            val checkmarkView = dialogItemLayout.findViewById<ImageView>(R.id.dialog_item_checkmark)

            checkmarkView.visibility = VISIBLE
            dialogItemLayout.setBackgroundResource(R.drawable.rounded_overlay)
        }
    }

    /**
     * Called when value changes. Updates summary, closes the [.dialog], an notifies on change listeners
     */
    private fun onChange() {
        summary = value

        dialog!!.cancel()

        if (mOnChangeListener != null) {
            mOnChangeListener!!.onPreferenceChange(this, value)
        }
    }

    private class SavedState : BaseSavedState {

        internal var value: String? = null
        internal var secondaryValue: Any? = null

        internal constructor(source: Parcel) : super(source) {
            value = source.readString()
            secondaryValue = source.readLong()
        }

        internal constructor(superState: Parcelable) : super(superState) {}

        override fun writeToParcel(dest: Parcel, flags: Int) {
            super.writeToParcel(dest, flags)
            dest.writeString(value)
        }

        companion object {
            @JvmField
            val CREATOR: Parcelable.Creator<SavedState> = object : Parcelable.Creator<SavedState> {
                override fun createFromParcel(`in`: Parcel): SavedState {
                    return SavedState(`in`)
                }

                override fun newArray(size: Int): Array<SavedState?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }
}


